evidence.go

```diff
@@ -34,139 +34,163 @@ type JSONCollection struct {
+	} else if e.PlatformClaims == nil {
+		return errors.New("missing platform claims")
+	} else if e.RealmClaims == nil {
+		return errors.New("missing realm claims")
 	}
-	if r == nil {
-		return errors.New("unmarshaling CCA claims: missing realm claims")
+
+	if err := e.PlatformClaims.Validate(); err != nil {
+		return fmt.Errorf("validation of cca-platform-claims failed: %w", err)
 	}
 
-	if err := e.SetClaims(p, r); err != nil {
-		return fmt.Errorf("setting claims: %w", err)
+	if err := e.RealmClaims.Validate(); err != nil {
+		return fmt.Errorf("validation of cca-realm-claims failed: %w", err)
 	}
 
 	return nil
 }
 
-func (e *Evidence) UnmarshalUnvalidatedJSON(data []byte) error {
-	p, r, err := e.doUnmarshalJSON(data)
+// UnmarshalCBOR extracts and validates the realm and platform tokens from the
+// serialized collection.
+func (e *Evidence) UnmarshalCBOR(buf []byte) error {
+	e.collection = &CBORCollection{}
+
+	err := dm.Unmarshal(buf, e.collection)
 	if err != nil {
-		return err
+		return fmt.Errorf("CBOR decoding of CCA evidence failed: %w", err)
 	}
 
-	if err := e.SetUnvalidatedClaims(p, r); err != nil {
-		return fmt.Errorf("setting claims: %w", err)
+	if e.collection.PlatformToken == nil {
+		return fmt.Errorf("CCA platform token not set")
+	}
+
+	if e.collection.RealmToken == nil {
+		return fmt.Errorf("CCA realm token not set")
+	}
+
+	// This will decode both platform and realm claims
+	err = e.decodeClaimsFromCBOR()
+	if err != nil {
+		return fmt.Errorf("decoding of CCA evidence failed: %w", err)
 	}
 
 	return nil
 }
 
-func (e *Evidence) doUnmarshalJSON(data []byte) (platform.IClaims, realm.IClaims, error) {
-	var c map[string]json.RawMessage
-
-	if err := json.Unmarshal(data, &c); err != nil {
-		return nil, nil, fmt.Errorf("unmarshaling CCA claims: %w", err)
+func (e *Evidence) MarshalJSON() ([]byte, error) {
+	pj, err := platform.EncodeClaimsToJSON(e.PlatformClaims)
+	if err != nil {
+		return nil, fmt.Errorf("error serializing platform claims: %w", err)
 	}
 
-	// platform
-	var p platform.IClaims
-	platToken, ok := c["cca-platform-token"]
-	if ok {
-		p = platform.NewClaims()
+	rj, err := realm.EncodeClaimsToJSON(e.RealmClaims)
+	if err != nil {
+		return nil, fmt.Errorf("error serializing realm claims: %w", err)
+	}
 
-		if err := json.Unmarshal(platToken, &p); err != nil {
-			return nil, nil, fmt.Errorf("unmarshaling platform claims: %w", err)
-		}
+	c := JSONCollection{
+		PlatformToken: pj,
+		RealmToken:    rj,
 	}
 
-	// realm
-	var r realm.IClaims
-	realmToken, ok := c["cca-realm-delegated-token"]
-	if ok {
-		r = realm.NewClaims()
+	return json.Marshal(c)
+}
 
-		if err := json.Unmarshal(realmToken, &r); err != nil {
-			return nil, nil, fmt.Errorf("unmarshaling realm claims: %w", err)
-		}
+func (e *Evidence) UnmarshalJSON(data []byte) error {
+	p, r, err := e.doUnmarshalJSON(data)
+	if err != nil {
+		return err
 	}
 
-	return p, r, nil
+	e.SetUnvalidatedClaims(p, r)
```

Comment:
Question: Why SetUnvalidatedClaims should be exposed outside?

Reply:
Because we should not be assuming what the calling code itends to do. If it knows the claims are invalid and still wants to set them, we should let it. (e.g. it may want to partially intialise the evidence in one place, and then complete initialisation later.)

(also note that `SetUnvalidateClaims` is already an exported API -- it is not being exported by this commit).